File: modules/editor/src/main/typescript/services/expression/JsonataEvaluator.test.ts
import { describe, it, expect, beforeEach } from "vitest";
import { JsonataEvaluator } from "./JsonataEvaluator";

describe("JsonataEvaluator", () => {
  let evaluator: JsonataEvaluator;

  beforeEach(() => {
    evaluator = new JsonataEvaluator();
  });

  describe("properties", () => {
    it("has correct type", () => {
      expect(evaluator.type).toBe("jsonata");
    });

    it("has correct name", () => {
      expect(evaluator.name).toBe("JSONata");
    });

    it("is marked as sandboxed", () => {
      expect(evaluator.isSandboxed).toBe(true);
    });
  });

  describe("initialize", () => {
    it("completes without error", async () => {
      await expect(evaluator.initialize()).resolves.toBeUndefined();
    });
  });

  describe("dispose", () => {
    it("completes without error", () => {
      expect(() => evaluator.dispose()).not.toThrow();
    });
  });

  describe("evaluate", () => {
    it("returns undefined for empty expression", async () => {
      const result = await evaluator.evaluate("", {});

      expect(result.success).toBe(true);
      expect(result.value).toBeUndefined();
    });

    it("returns undefined for whitespace-only expression", async () => {
      const result = await evaluator.evaluate("   ", {});

      expect(result.success).toBe(true);
      expect(result.value).toBeUndefined();
    });

    it("evaluates simple property access", async () => {
      const context = { name: "John" };

      const result = await evaluator.evaluate("name", context);

      expect(result.success).toBe(true);
      expect(result.value).toBe("John");
    });

    it("evaluates nested property access", async () => {
      const context = { user: { name: "John", age: 30 } };

      const result = await evaluator.evaluate("user.name", context);

      expect(result.success).toBe(true);
      expect(result.value).toBe("John");
    });

    it("evaluates array access", async () => {
      const context = { items: ["a", "b", "c"] };

      const result = await evaluator.evaluate("items[0]", context);

      expect(result.success).toBe(true);
      expect(result.value).toBe("a");
    });

    it("evaluates array mapping", async () => {
      const context = { items: [{ price: 10 }, { price: 20 }, { price: 30 }] };

      const result = await evaluator.evaluate("items.price", context);

      expect(result.success).toBe(true);
      // JSONata returns arrays with a 'sequence' property, so we check values
      expect([...result.value]).toEqual([10, 20, 30]);
    });

    it("evaluates array filtering", async () => {
      const context = { items: [{ active: true }, { active: false }, { active: true }] };

      const result = await evaluator.evaluate("items[active].active", context);

      expect(result.success).toBe(true);
      // JSONata returns arrays with a 'sequence' property, so we check values
      expect([...result.value]).toEqual([true, true]);
    });

    it("evaluates arithmetic expressions", async () => {
      const context = { a: 10, b: 5 };

      const result = await evaluator.evaluate("a + b", context);

      expect(result.success).toBe(true);
      expect(result.value).toBe(15);
    });

    it("evaluates string concatenation", async () => {
      const context = { first: "John", last: "Doe" };

      const result = await evaluator.evaluate('first & " " & last', context);

      expect(result.success).toBe(true);
      expect(result.value).toBe("John Doe");
    });

    it("evaluates conditional expression", async () => {
      const context = { active: true };

      const result = await evaluator.evaluate('active ? "Yes" : "No"', context);

      expect(result.success).toBe(true);
      expect(result.value).toBe("Yes");
    });

    it("evaluates $sum aggregation", async () => {
      const context = { items: [{ price: 10 }, { price: 20 }, { price: 30 }] };

      const result = await evaluator.evaluate("$sum(items.price)", context);

      expect(result.success).toBe(true);
      expect(result.value).toBe(60);
    });

    it("evaluates $count aggregation", async () => {
      const context = { items: [1, 2, 3, 4, 5] };

      const result = await evaluator.evaluate("$count(items)", context);

      expect(result.success).toBe(true);
      expect(result.value).toBe(5);
    });

    it("evaluates $average aggregation", async () => {
      const context = { values: [10, 20, 30] };

      const result = await evaluator.evaluate("$average(values)", context);

      expect(result.success).toBe(true);
      expect(result.value).toBe(20);
    });

    it("returns undefined for missing property", async () => {
      const context = { name: "John" };

      const result = await evaluator.evaluate("missing", context);

      expect(result.success).toBe(true);
      expect(result.value).toBeUndefined();
    });

    it("returns error for invalid syntax", async () => {
      const result = await evaluator.evaluate("invalid[[[", {});

      expect(result.success).toBe(false);
      expect(result.error).toBeDefined();
    });

    it("returns error for invalid function call", async () => {
      const result = await evaluator.evaluate("$unknownFunction()", {});

      expect(result.success).toBe(false);
      expect(result.error).toBeDefined();
    });

    it("handles complex nested expressions", async () => {
      const context = {
        order: {
          items: [
            { name: "Item 1", price: 100, quantity: 2 },
            { name: "Item 2", price: 50, quantity: 3 },
          ],
        },
      };

      const result = await evaluator.evaluate(
        "$sum(order.items.(price * quantity))",
        context
      );

      expect(result.success).toBe(true);
      expect(result.value).toBe(350);
    });

    it("handles empty context", async () => {
      const result = await evaluator.evaluate("$now()", {});

      expect(result.success).toBe(true);
      // $now() returns a timestamp
      expect(typeof result.value).toBe("string");
    });
  });
});

File: modules/generation/src/test/kotlin/app/epistola/generation/expression/JsonataEvaluatorTest.kt
package app.epistola.generation.expression

import java.util.Locale
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

class JsonataEvaluatorTest {
    private val evaluator = JsonataEvaluator()
    private val dutchEvaluator = JsonataEvaluator(locale = Locale.forLanguageTag("nl-NL"))
    private val germanEvaluator = JsonataEvaluator(locale = Locale.forLanguageTag("de-DE"))

    @Test
    fun `evaluate simple path`() {
        val data = mapOf("name" to "John")
        assertEquals("John", evaluator.evaluate("name", data))
    }

    @Test
    fun `evaluate nested path`() {
        val data = mapOf(
            "customer" to mapOf(
                "name" to "Alice",
                "address" to mapOf("city" to "Amsterdam"),
            ),
        )
        assertEquals("Alice", evaluator.evaluate("customer.name", data))
        assertEquals("Amsterdam", evaluator.evaluate("customer.address.city", data))
    }

    @Test
    fun `evaluate array access`() {
        val data = mapOf(
            "items" to listOf("first", "second", "third"),
        )
        assertEquals("first", evaluator.evaluate("items[0]", data))
        assertEquals("second", evaluator.evaluate("items[1]", data))
    }

    @Test
    fun `evaluate array mapping`() {
        val data = mapOf(
            "products" to listOf(
                mapOf("price" to 10),
                mapOf("price" to 20),
                mapOf("price" to 30),
            ),
        )
        // JSONata: products.price returns an array of prices
        val result = evaluator.evaluate("products.price", data)
        assertEquals(listOf(10, 20, 30), result)
    }

    @Test
    fun `evaluate sum aggregation`() {
        val data = mapOf(
            "items" to listOf(
                mapOf("price" to 10),
                mapOf("price" to 20),
                mapOf("price" to 30),
            ),
        )
        assertEquals(60, evaluator.evaluate("\$sum(items.price)", data))
    }

    @Test
    fun `evaluate string concatenation`() {
        val data = mapOf("first" to "John", "last" to "Doe")
        assertEquals("John Doe", evaluator.evaluate("first & \" \" & last", data))
    }

    @Test
    fun `evaluate conditional expression`() {
        val data = mapOf("active" to true)
        assertEquals("Yes", evaluator.evaluate("active ? \"Yes\" : \"No\"", data))

        val data2 = mapOf("active" to false)
        assertEquals("No", evaluator.evaluate("active ? \"Yes\" : \"No\"", data2))
    }

    @Test
    fun `evaluate filter expression`() {
        val data = mapOf(
            "items" to listOf(
                mapOf("name" to "A", "active" to true),
                mapOf("name" to "B", "active" to false),
                mapOf("name" to "C", "active" to true),
            ),
        )
        // JSONata: items[active] filters to only active items
        val result = evaluator.evaluate("items[active]", data) as List<*>
        assertEquals(2, result.size)
    }

    @Test
    fun `evaluate returns null for missing path`() {
        val data = mapOf("name" to "John")
        assertNull(evaluator.evaluate("missing", data))
    }

    @Test
    fun `evaluate returns null for empty expression`() {
        val data = mapOf("name" to "John")
        assertNull(evaluator.evaluate("", data))
        assertNull(evaluator.evaluate("  ", data))
    }

    @Test
    fun `evaluateToString formats values correctly`() {
        val data = mapOf(
            "string" to "hello",
            "number" to 42,
            "decimal" to 3.14,
            "boolean" to true,
        )
        assertEquals("hello", evaluator.evaluateToString("string", data))
        assertEquals("42", evaluator.evaluateToString("number", data))
        assertEquals("3.14", evaluator.evaluateToString("decimal", data))
        assertEquals("true", evaluator.evaluateToString("boolean", data))
        assertEquals("", evaluator.evaluateToString("missing", data))
    }

    @Test
    fun `evaluateCondition with truthy values`() {
        val data = mapOf(
            "trueBoolean" to true,
            "nonEmptyString" to "hello",
            "nonZeroNumber" to 42,
        )
        assertTrue(evaluator.evaluateCondition("trueBoolean", data))
        assertTrue(evaluator.evaluateCondition("nonEmptyString", data))
        assertTrue(evaluator.evaluateCondition("nonZeroNumber", data))
    }

    @Test
    fun `evaluateCondition with falsy values`() {
        val data = mapOf(
            "falseBoolean" to false,
            "emptyString" to "",
            "zero" to 0,
        )
        assertFalse(evaluator.evaluateCondition("falseBoolean", data))
        assertFalse(evaluator.evaluateCondition("emptyString", data))
        assertFalse(evaluator.evaluateCondition("zero", data))
        assertFalse(evaluator.evaluateCondition("missing", data))
    }

    @Test
    fun `evaluateIterable returns list`() {
        val data = mapOf(
            "items" to listOf("a", "b", "c"),
        )
        assertEquals(listOf("a", "b", "c"), evaluator.evaluateIterable("items", data))
        assertEquals(emptyList<Any?>(), evaluator.evaluateIterable("missing", data))
    }

    @Test
    fun `processTemplate replaces expressions`() {
        val data = mapOf(
            "name" to "John",
            "company" to "Acme",
        )
        assertEquals(
            "Hello John from Acme!",
            evaluator.processTemplate("Hello {{name}} from {{company}}!", data),
        )
    }

    @Test
    fun `evaluate with loop context`() {
        val data = mapOf("global" to "value")
        val loopContext = mapOf("item" to mapOf("name" to "Product"))
        assertEquals("Product", evaluator.evaluate("item.name", data, loopContext))
        assertEquals("value", evaluator.evaluate("global", data, loopContext))
    }

    @Test
    fun `loop context overrides data`() {
        val data = mapOf("item" to mapOf("name" to "DataValue"))
        val loopContext = mapOf("item" to mapOf("name" to "LoopValue"))
        assertEquals("LoopValue", evaluator.evaluate("item.name", data, loopContext))
    }

    // --- $formatDate custom function ---

    @Test
    fun `formatDate with dd-MM-yyyy`() {
        val data = mapOf("d" to "2024-01-15")
        assertEquals("15-01-2024", evaluator.evaluate("\$formatDate(d, 'dd-MM-yyyy')", data))
    }

    @Test
    fun `formatDate with dd slash MM slash yyyy`() {
        val data = mapOf("d" to "2024-01-15")
        assertEquals("15/01/2024", evaluator.evaluate("\$formatDate(d, 'dd/MM/yyyy')", data))
    }

    @Test
    fun `formatDate with d MMMM yyyy`() {
        val data = mapOf("d" to "2024-01-15")
        assertEquals("15 January 2024", evaluator.evaluate("\$formatDate(d, 'd MMMM yyyy')", data))
    }

    @Test
    fun `formatDate with d MMMM yyyy no leading zero`() {
        val data = mapOf("d" to "2024-07-05")
        assertEquals("5 July 2024", evaluator.evaluate("\$formatDate(d, 'd MMMM yyyy')", data))
    }

    @Test
    fun `formatDate with invalid date returns raw value`() {
        val data = mapOf("d" to "not-a-date")
        assertEquals("not-a-date", evaluator.evaluate("\$formatDate(d, 'dd-MM-yyyy')", data))
    }

    @Test
    fun `formatDate in string concatenation`() {
        val data = mapOf("dueDate" to "2024-02-15")
        assertEquals(
            "Due: 15-02-2024",
            evaluator.evaluate("\"Due: \" & \$formatDate(dueDate, 'dd-MM-yyyy')", data),
        )
    }

    @Test
    fun `formatDate with missing field returns null`() {
        val data = mapOf("other" to "2024-01-15")
        assertNull(evaluator.evaluate("\$formatDate(missing, 'dd-MM-yyyy')", data))
    }

    // --- $formatDate with datetimes ---

    @Test
    fun `formatDate with local datetime`() {
        val data = mapOf("ts" to "2024-01-15T14:30:00")
        assertEquals(
            "15-01-2024 14:30",
            evaluator.evaluate("\$formatDate(ts, 'dd-MM-yyyy HH:mm')", data),
        )
    }

    @Test
    fun `formatDate with UTC datetime converts to Amsterdam timezone`() {
        // UTC 12:00 → Amsterdam CET is UTC+1 → 13:00
        val data = mapOf("ts" to "2024-01-15T12:00:00Z")
        assertEquals(
            "15-01-2024 13:00",
            evaluator.evaluate("\$formatDate(ts, 'dd-MM-yyyy HH:mm')", data),
        )
    }

    @Test
    fun `formatDate with offset datetime converts to Amsterdam timezone`() {
        // +05:00 at 17:00 → UTC 12:00 → Amsterdam CET (UTC+1) → 13:00
        val data = mapOf("ts" to "2024-01-15T17:00:00+05:00")
        assertEquals(
            "15-01-2024 13:00",
            evaluator.evaluate("\$formatDate(ts, 'dd-MM-yyyy HH:mm')", data),
        )
    }

    @Test
    fun `formatDate with UTC datetime and date-only pattern`() {
        val data = mapOf("ts" to "2024-01-15T14:30:00Z")
        assertEquals(
            "15-01-2024",
            evaluator.evaluate("\$formatDate(ts, 'dd-MM-yyyy')", data),
        )
    }

    @Test
    fun `formatDate with datetime including seconds`() {
        val data = mapOf("ts" to "2024-01-15T14:30:45")
        assertEquals(
            "14:30:45",
            evaluator.evaluate("\$formatDate(ts, 'HH:mm:ss')", data),
        )
    }

    // --- $formatDate locale awareness (month name spelling) ---

    @Test
    fun `formatDate localizes the month name for nl-NL`() {
        // Dutch month names are lowercase in CLDR.
        val data = mapOf("d" to "2024-01-15")
        assertEquals("15 januari 2024", dutchEvaluator.evaluate("\$formatDate(d, 'd MMMM yyyy')", data))
    }

    @Test
    fun `formatDate localizes the month name for de-DE`() {
        val data = mapOf("d" to "2024-03-15")
        assertEquals("15 März 2024", germanEvaluator.evaluate("\$formatDate(d, 'd MMMM yyyy')", data))
    }

    @Test
    fun `formatDate numeric tokens are locale-agnostic`() {
        // yyyy/MM/dd render identically regardless of locale — only name tokens localize.
        val data = mapOf("d" to "2026-04-04")
        assertEquals("2026-04-04", dutchEvaluator.evaluate("\$formatDate(d, 'yyyy-MM-dd')", data))
    }

    // --- $formatLocaleNumber custom function ---

    @Test
    fun `formatLocaleNumber uses en-US separators by default`() {
        val data = mapOf("amount" to 1234.56)
        assertEquals("1,234.56", evaluator.evaluate("\$formatLocaleNumber(amount, '#,##0.00')", data))
    }

    @Test
    fun `formatLocaleNumber uses nl-NL separators`() {
        val data = mapOf("amount" to 1234.56)
        assertEquals("1.234,56", dutchEvaluator.evaluate("\$formatLocaleNumber(amount, '#,##0.00')", data))
    }

    @Test
    fun `formatLocaleNumber uses de-DE separators`() {
        val data = mapOf("amount" to 1234.56)
        assertEquals("1.234,56", germanEvaluator.evaluate("\$formatLocaleNumber(amount, '#,##0.00')", data))
    }

    @Test
    fun `formatLocaleNumber whole number with grouping`() {
        val data = mapOf("n" to 1234)
        assertEquals("1.234", dutchEvaluator.evaluate("\$formatLocaleNumber(n, '#,##0')", data))
    }

    @Test
    fun `formatLocaleNumber percent multiplies by 100 and localizes the decimal`() {
        val data = mapOf("rate" to 0.21)
        assertEquals("21,0%", dutchEvaluator.evaluate("\$formatLocaleNumber(rate, '0.0%')", data))
        assertEquals("21.0%", evaluator.evaluate("\$formatLocaleNumber(rate, '0.0%')", data))
    }

    @Test
    fun `formatLocaleNumber honours the explicit negative subpattern`() {
        val data = mapOf("n" to -12.5)
        assertEquals(
            "(12,50)",
            dutchEvaluator.evaluate("\$formatLocaleNumber(n, '#,##0.00;(#,##0.00)')", data),
        )
    }

    @Test
    fun `formatLocaleNumber rounds HALF_EVEN (banker's) by default`() {
        // DecimalFormat's default rounding is HALF_EVEN: ties go to the nearest
        // even digit. The editor preview pins Intl roundingMode halfEven to match.
        assertEquals("8", evaluator.evaluate("\$formatLocaleNumber(n, '#,##0')", mapOf("n" to 8.5)))
        assertEquals("10", evaluator.evaluate("\$formatLocaleNumber(n, '#,##0')", mapOf("n" to 9.5)))
        assertEquals("2", evaluator.evaluate("\$formatLocaleNumber(n, '#,##0')", mapOf("n" to 2.5)))
    }

    @Test
    fun `formatLocaleNumber with non-numeric value returns the raw value`() {
        val data = mapOf("n" to "not-a-number")
        assertEquals("not-a-number", evaluator.evaluate("\$formatLocaleNumber(n, '#,##0.00')", data))
    }

    @Test
    fun `formatLocaleNumber with missing field returns null`() {
        val data = mapOf("other" to 1)
        assertNull(evaluator.evaluate("\$formatLocaleNumber(missing, '#,##0.00')", data))
    }
}
